consensus/consensus.go

```diff
@@ -80,10 +80,17 @@ type Engine interface {
 	Prepare(chain ChainReader, header *types.Header) error
 
 	// Finalize runs any post-transaction state modifications (e.g. block rewards)
-	// and assembles the final block.
+	// but does not assemble the block.
 	// Note: The block header and state database might be updated to reflect any
 	// consensus rules that happen at finalization (e.g. block rewards).
 	Finalize(chain ChainReader, header *types.Header, state *state.StateDB, txs []*types.Transaction,
+		uncles []*types.Header)
+
+	// FinalizeAndAssemble runs any post-transaction state modifications (e.g. block rewards)
+	// and assembles the final block.
+	// Note: The block header and state database might be updated to reflect any
+	// consensus rules that happen at finalization (e.g. block rewards).
+	FinalizeAndAssemble(chain ChainReader, header *types.Header, state *state.StateDB, txs []*types.Transaction,
```

Comment:
Hmm, wondering if we need this at all. This is essentially `Finalize` + `types.NewBlock(header, txs, nil, receipts)`, isn't it? If yes, I think API wise it would be cleaner to have Finalize **not** create and return the block and then outside code can do if needed.

Reply:
I could do that, but right now the Clique block assembly makes sure that the uncles are ignored. If we move block creation outside, then we won't have that guarantee.

Reply:
:+1: